### PR TITLE
fix(storage): fix some delete range related issues in merged imm

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

fix some delete range related issues in merged imms
View comments in https://github.com/risingwavelabs/risingwave/pull/7368 for details.

## Checklist For Contributors

- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by StrikeW** [comment]
Can you also enhance the unit test `test_merge_imms_delete_range` to cover the buggy use case.

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs [hunk 1]
```diff
@@ -531,6 +532,10 @@ impl SharedBufferBatch {
         self.inner.range_tombstone_list.clone()
     }
 
+    pub fn get_delete_range_tombstones_for_query(&self) -> Vec<DeleteRangeTombstone> {
```
Comment (StrikeW) [inline src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:535]: Seems the above `get_delete_range_tombstones` is reserved for its using in `compact_shared_buffer`. Can we unify `range_tombstone_list` and `range_tombstone_list_for_query` to a single one? (Maybe in a follow-up refactor PR)

**Review by StrikeW — COMMENTED** [review]
Rest lgtm, PTAL @hzxa21 @Little-Wallace

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs (hunk 1 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:535]: > Seems the above `get_delete_range_tombstones` is reserved for its using in `compact_shared_buffer`. Can we unify `range_tombstone_list` and `range_tombstone_list_for_query` to a single one? (Maybe in a follow-up refactor PR)

Can be done later.
5aee4b37dff3c75269b0e7835791dcd0

Comment (StrikeW) [reply src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:535]: > > Seems the above `get_delete_range_tombstones` is reserved for its using in `compact_shared_buffer`. Can we unify `range_tombstone_list` and `range_tombstone_list_for_query` to a single one? (Maybe in a follow-up refactor PR)
> 
> Can be done later. 5aee4b37dff3c75269b0e7835791dcd0

typo? Should be a link to some PR or issue?

File: src/storage/src/hummock/compactor/shared_buffer_compact.rs [hunk 2]
```diff
@@ -310,9 +327,12 @@ pub async fn merge_imms_in_memory(
         inner: Arc::new(SharedBufferBatchInner::new_with_multi_epoch_batches(
             epochs,
             merged_payload,
+            smallest_table_key,
+            largest_table_key,
             kv_count,
             merged_imm_ids,
             range_tombstone_list,
+            collector.get_all_tombstones(),
```
Comment (Little-Wallace) [inline src/storage/src/hummock/compactor/shared_buffer_compact.rs:335]: Do we still need to keep origin `range_tombstone_list` ?

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs (hunk 1 above)
Comment (StrikeW) [reply src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:535]: I found that there is a PR (https://github.com/risingwavelabs/risingwave/pull/8927) to optimize the delete range implementation overall. Do you plan to unify the `range_tombstone_list` in that PR? @soundOfDestiny

Comment (soundOfDestiny) [reply src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:535]: > I found that there is a PR (#8927) to optimize the delete range implementation overall. Do you plan to unify the two `range_tombstone_list` in that PR? @soundOfDestiny

I have also written a #8927 version of this PR.

Comment (StrikeW) [reply src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:535]: Ok, we can merge current PR as a quick fix of the bug

**Review by StrikeW — APPROVED** [review]
We can merge the PR as a quick fix for the bug. Leaving the refactor of `range_tombstone_list` to #8927

**Review by Little-Wallace — APPROVED** [review]
rest LGTM

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs (hunk 1 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:535]: > Seems the above `get_delete_range_tombstones` is reserved for its using in `compact_shared_buffer`. Can we unify `range_tombstone_list` and `range_tombstone_list_for_query` to a single one? (Maybe in a follow-up refactor PR)

fixed in #9255

File: src/storage/src/hummock/compactor/shared_buffer_compact.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/compactor/shared_buffer_compact.rs:335]: > Do we still need to keep origin `range_tombstone_list` ?

fixed in #9255

**Comment by soundOfDestiny** [comment]
> We can merge the PR as a quick fix for the bug. Leaving the refactor of `range_tombstone_list` to #8927

#9255 has done dedup :-)

**Pull request merged**